lib/execution_engine2/authorization/authstrategy.py

```diff
@@ -0,0 +1,139 @@
+    :param job: a Job model object
+    :param user_id: string - the user id
+    :param token: string - the user's auth token
+    :param config: dict - the service config
+    :returns: List[bool] - Has True values if the user can write job info, False otherwise
+    """
+    return _check_permissions_list(jobs, user_id, token, config, level="write")
+
+
+def _check_permissions(job: Job, user_id: str, token: str, config: Dict[str, str], level="read") -> bool:
+    """
+    Returns a job permissions, for either read or write ability
+    :param job: a Job model object
+    :param user_id: string - the user id
+    :param token: string - the user's auth token
+    :param config: dict - the service config
+    :param level: string - if "read", then returns the read value, if "write", return whether the user can write.
+    :returns: bool - True if the permission is valid, False otherwise.
+    """
+    if user_id == job.user:
+        return True
+    if job.authstrat == KBASE_WS_AUTHSTRAT:
+        ws_auth = WorkspaceAuth(token, user_id, config['workspace-url'])
+        if level == "read":
+            return ws_auth.can_read(job.wsid)
+        else:
+            return ws_auth.can_write(job.wsid)
+    else:
+        return user_id == job.user
+
+
+def _check_permissions_list(jobs: List[Job], user_id: str, token: str, config: Dict[str, str], level="read") -> bool:
+    """
+    Returns True for each job the user has read access to, and False for the ones they don't.
+    :param job: a Job model object
+    :param user_id: string - the user id
+    :param token: string - the user's auth token
+    :param config: dict - the service config
+    :param level: string - if "read" then tests if the Job can be read, otherwise checks if it can be written
+    :returns: List[bool] - Has True values if the user can write job info, False otherwise
+    """
+
+    # This is a little tricky, but makes either 0 or 1 workspace call.
+    # 0. Init list of permissions to all be False.
+    # 1. Go over the list of jobs. The ones that match user ids are ok, so set them as True.
+    # 2. The ones that don't match, and have an authstrat of kbaseworkspace, get their ws ids and
+    #    put them aside in a separate dict. This should have a key of wsid and a (growing) list
+    #    of job indices as values.
+    # 3. The rest (not the same user, and a different authstrat) are not allowed, so leave them
+    #    marked False in the perms list
+    # 4. Now, if there's any ids in that wsid dict, use WorkspaceAuth to look 'em up.
+    # 5. We can now unravel the mappings. ws_auth_perms -> wsid -> list of indices of jobs with that
+    #    wsid.
+    # should be just a single ws call to get_permissions_mass, and up to 2 loops over the list of
+    # jobs (O(2N) if there's a unique ws id for each job, and fewer comparisons)
+
+    perms = len(jobs) * [False]    # permissions start False
```

Comment:
The alternative could be to use a `defaultdict(lambda: False)` instead of this list

Reply:
or `defaultdict(bool)`?

Reply:
That's the right thing to do. Dangit, I always forget about defaultdict...

Reply:
Oh, I thought you meant below where it checks for dict initialization. That list needs to be a fixed size, matching the number of jobs.

Reply:
Updated to defaultdict